File: backend/services/user_service.py
import json
import os
from datetime import datetime, date
from typing import Dict, List, Optional
from models.user_models import (
    UserProfile, NotificationSettings, HealthPreferences, PrivacySettings,
    DailyGoal, CreateDailyGoal, UpdateDailyGoal, UpdateNotificationSettings,
    UpdateHealthPreferences, UpdatePrivacySettings, GoalType
)
import uuid

class UserService:
    def __init__(self, data_dir: str = "data"):
        self.data_dir = data_dir
        os.makedirs(data_dir, exist_ok=True)
        self.users_file = os.path.join(data_dir, "users.json")
        self.drinks_file = os.path.join(data_dir, "drinks.json")
        self._load_data()

    def _load_data(self):
        """Load user data from file"""
        try:
            with open(self.users_file, 'r') as f:
                self.users_data = json.load(f)
        except FileNotFoundError:
            self.users_data = {}

    def _save_data(self):
        """Save user data to file"""
        with open(self.users_file, 'w') as f:
            json.dump(self.users_data, f, indent=2, default=str)

    def get_or_create_user(self, user_id: str = "default") -> UserProfile:
        """Get user profile or create default one"""
        if user_id not in self.users_data:
            # Create default user profile
            now = datetime.now()
            default_goals = [
                DailyGoal(
                    id=str(uuid.uuid4()),
                    name="Daily Calories",
                    target=2000,
                    unit="kcal",
                    type=GoalType.calories,
                    created_at=now,
                    current=0
                ),
                DailyGoal(
                    id=str(uuid.uuid4()),
                    name="Daily Water",
                    target=2000,
                    unit="ml",
                    type=GoalType.water,
                    created_at=now,
                    current=0
                ),
                DailyGoal(
                    id=str(uuid.uuid4()),
                    name="Sugar Limit",
                    target=50,
                    unit="g",
                    type=GoalType.sugar,
                    created_at=now,
                    current=0
                )
            ]
            
            user_profile = UserProfile(
                user_id=user_id,
                daily_goals=default_goals,
                created_at=now,
                updated_at=now
            )
            self.users_data[user_id] = user_profile.dict()
            self._save_data()
        
        user_data = self.users_data[user_id]
        return UserProfile(**user_data)

    def update_notifications(self, user_id: str, settings: UpdateNotificationSettings) -> NotificationSettings:
        """Update notification settings"""
        user = self.get_or_create_user(user_id)
        
        # Update only provided fields
        update_data = settings.dict(exclude_unset=True)
        for key, value in update_data.items():
            setattr(user.notifications, key, value)
        
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return user.notifications

    def update_health_preferences(self, user_id: str, preferences: UpdateHealthPreferences) -> HealthPreferences:
        """Update health preferences"""
        user = self.get_or_create_user(user_id)
        
        # Update only provided fields
        update_data = preferences.dict(exclude_unset=True)
        for key, value in update_data.items():
            setattr(user.health_preferences, key, value)
        
        # Auto-calculate target calories based on health data
        if user.health_preferences.age and user.health_preferences.weight and user.health_preferences.height:
            target_calories = self._calculate_target_calories(user.health_preferences)
            user.health_preferences.target_calories = target_calories
        
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return user.health_preferences

    def update_privacy_settings(self, user_id: str, settings: UpdatePrivacySettings) -> PrivacySettings:
        """Update privacy settings"""
        user = self.get_or_create_user(user_id)
        
        # Update only provided fields
        update_data = settings.dict(exclude_unset=True)
        for key, value in update_data.items():
            setattr(user.privacy_settings, key, value)
        
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return user.privacy_settings

    def create_daily_goal(self, user_id: str, goal_data: CreateDailyGoal) -> DailyGoal:
        """Create a new daily goal"""
        user = self.get_or_create_user(user_id)
        
        new_goal = DailyGoal(
            id=str(uuid.uuid4()),
            name=goal_data.name,
            target=goal_data.target,
            unit=goal_data.unit,
            type=goal_data.type,
            created_at=datetime.now(),
            current=0
        )
        
        user.daily_goals.append(new_goal)
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return new_goal

    def update_daily_goal(self, user_id: str, goal_id: str, goal_update: UpdateDailyGoal) -> DailyGoal:
        """Update an existing daily goal"""
        user = self.get_or_create_user(user_id)
        
        goal_index = next((i for i, g in enumerate(user.daily_goals) if g.id == goal_id), None)
        if goal_index is None:
            raise ValueError(f"Goal with id {goal_id} not found")
        
        goal = user.daily_goals[goal_index]
        update_data = goal_update.dict(exclude_unset=True)
        
        for key, value in update_data.items():
            setattr(goal, key, value)
        
        # Check if goal is achieved
        goal.is_achieved = goal.current >= goal.target
        
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return goal

    def get_daily_goals(self, user_id: str) -> List[DailyGoal]:
        """Get all daily goals for user"""
        user = self.get_or_create_user(user_id)
        return user.daily_goals

    def update_goals_from_drink(self, user_id: str, nutrition_data: Dict):
        """Update daily goals based on consumed drink"""
        user = self.get_or_create_user(user_id)
        
        # Reset daily goals if it's a new day
        self._reset_daily_goals_if_new_day(user)
        
        # Update each goal based on nutrition data
        for goal in user.daily_goals:
            if goal.type == GoalType.calories:
                goal.current += nutrition_data.get('calories', 0)
            elif goal.type == GoalType.sugar:
                goal.current += nutrition_data.get('sugar_g', 0)
            elif goal.type == GoalType.caffeine:
                goal.current += nutrition_data.get('caffeine_mg', 0)
            elif goal.type == GoalType.water:
                goal.current += nutrition_data.get('water_ml', 0)
            elif goal.type == GoalType.sodium:
                goal.current += nutrition_data.get('sodium_mg', 0)
            
            # Check if goal is achieved
            goal.is_achieved = goal.current >= goal.target
        
        user.updated_at = datetime.now()
        self.users_data[user_id] = user.dict()
        self._save_data()
        
        return user.daily_goals

    def _reset_daily_goals_if_new_day(self, user: UserProfile):
        """Reset daily goal progress if it's a new day"""
        today = date.today()
        last_update = user.updated_at.date() if isinstance(user.updated_at, datetime) else date.fromisoformat(user.updated_at)
        
        if today > last_update:
            for goal in user.daily_goals:
                goal.current = 0
                goal.is_achieved = False

    def _calculate_target_calories(self, health_prefs: HealthPreferences) -> int:
        """Calculate target daily calories using Mifflin-St Jeor equation"""
        if not all([health_prefs.age, health_prefs.weight, health_prefs.height]):
            return 2000  # Default
        
        # Assume male for simplicity (in real app, add gender field)
        bmr = 10 * health_prefs.weight + 6.25 * health_prefs.height - 5 * health_prefs.age + 5
        
        # Activity multipliers
        activity_multipliers = {
            "sedentary": 1.2,
            "light": 1.375,
            "moderate": 1.55,
            "active": 1.725,
            "very_active": 1.9
        }
        
        multiplier = activity_multipliers.get(health_prefs.activity_level, 1.55)
        return int(bmr * multiplier)

    def get_achievements(self, user_id: str) -> List[Dict]:
        """Get user achievements based on goals"""
        user = self.get_or_create_user(user_id)
        achievements = []
        
        for goal in user.daily_goals:
            if goal.is_achieved:
                achievements.append({
                    "goal_name": goal.name,
                    "target": goal.target,
                    "current": goal.current,
                    "unit": goal.unit,
                    "achieved_at": datetime.now().isoformat()
                })
        
        return achievements

    def get_user_stats(self, user_id: str) -> Dict:
        """Get comprehensive user statistics"""
        user = self.get_or_create_user(user_id)
        
        total_goals = len(user.daily_goals)
        achieved_goals = sum(1 for goal in user.daily_goals if goal.is_achieved)
        
        return {
            "total_goals": total_goals,
            "achieved_goals": achieved_goals,
            "achievement_rate": achieved_goals / total_goals if total_goals > 0 else 0,
            "goals": [goal.dict() for goal in user.daily_goals],
            "health_preferences": user.health_preferences.dict(),
            "notifications": user.notifications.dict(),
            "privacy": user.privacy_settings.dict()
        }

File: client/src/services/userService.ts
const API_BASE_URL = 'http://localhost:8000';

interface NotificationSettings {
  daily_reminders: boolean;
  goal_achievements: boolean;
  health_tips: boolean;
  weekly_reports: boolean;
  reminder_time: string;
}

interface HealthPreferences {
  age?: number;
  weight?: number;
  height?: number;
  activity_level: string;
  dietary_restrictions?: string;
  health_goals?: string;
  target_calories?: number;
  target_water_ml?: number;
}

interface PrivacySettings {
  data_collection: boolean;
  analytics_tracking: boolean;
  personalized_ads: boolean;
  share_with_partners: boolean;
}

interface DailyGoal {
  id: string;
  name: string;
  target: number;
  current: number;
  unit: string;
  type: string;
  created_at: string;
  is_achieved: boolean;
}

export class UserService {
  private static userId = 'default';

  // Notifications
  static async getNotifications(): Promise<NotificationSettings> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/notifications`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available, using local settings');
    }
    
    // Fallback to localStorage
    const stored = localStorage.getItem('notifications');
    return stored ? JSON.parse(stored) : {
      daily_reminders: true,
      goal_achievements: true,
      health_tips: false,
      weekly_reports: true,
      reminder_time: "09:00"
    };
  }

  static async updateNotifications(settings: Partial<NotificationSettings>): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/notifications`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(settings)
      });
      if (response.ok) return;
    } catch (error) {
      console.warn('Backend not available, saving locally');
    }
    
    // Fallback to localStorage
    const current = await this.getNotifications();
    localStorage.setItem('notifications', JSON.stringify({ ...current, ...settings }));
  }

  // Health Preferences
  static async getHealthPreferences(): Promise<HealthPreferences> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/health-preferences`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available, using local preferences');
    }
    
    // Fallback to localStorage
    const stored = localStorage.getItem('healthPreferences');
    return stored ? JSON.parse(stored) : {
      activity_level: "moderate"
    };
  }

  static async updateHealthPreferences(preferences: Partial<HealthPreferences>): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/health-preferences`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(preferences)
      });
      if (response.ok) return;
    } catch (error) {
      console.warn('Backend not available, saving locally');
    }
    
    // Fallback to localStorage
    const current = await this.getHealthPreferences();
    localStorage.setItem('healthPreferences', JSON.stringify({ ...current, ...preferences }));
  }

  // Privacy Settings
  static async getPrivacySettings(): Promise<PrivacySettings> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/privacy`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available, using local settings');
    }
    
    // Fallback to localStorage
    const stored = localStorage.getItem('privacySettings');
    return stored ? JSON.parse(stored) : {
      data_collection: true,
      analytics_tracking: false,
      personalized_ads: false,
      share_with_partners: false
    };
  }

  static async updatePrivacySettings(settings: Partial<PrivacySettings>): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/privacy`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(settings)
      });
      if (response.ok) return;
    } catch (error) {
      console.warn('Backend not available, saving locally');
    }
    
    // Fallback to localStorage
    const current = await this.getPrivacySettings();
    localStorage.setItem('privacySettings', JSON.stringify({ ...current, ...settings }));
  }

  // Daily Goals
  static async getDailyGoals(): Promise<DailyGoal[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/daily-goals`);
      if (response.ok) {
        const data = await response.json();
        return data.goals;
      }
    } catch (error) {
      console.warn('Backend not available, using default goals');
    }
    
    // Fallback to default goals
    return [
      {
        id: '1',
        name: 'Daily Calories',
        target: 2000,
        current: 0,
        unit: 'kcal',
        type: 'calories',
        created_at: new Date().toISOString(),
        is_achieved: false
      },
      {
        id: '2',
        name: 'Daily Water',
        target: 2000,
        current: 0,
        unit: 'ml',
        type: 'water',
        created_at: new Date().toISOString(),
        is_achieved: false
      },
      {
        id: '3',
        name: 'Sugar Limit',
        target: 50,
        current: 0,
        unit: 'g',
        type: 'sugar',
        created_at: new Date().toISOString(),
        is_achieved: false
      }
    ];
  }

  static async createDailyGoal(goal: { name: string; target: number; unit: string; type: string }): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/daily-goals`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(goal)
      });
      if (response.ok) return;
    } catch (error) {
      console.warn('Backend not available, goal not synced');
    }
  }

  static async updateDailyGoal(goalId: string, updates: { target?: number; current?: number }): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/daily-goals/${goalId}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updates)
      });
      if (response.ok) return;
    } catch (error) {
      console.warn('Backend not available, goal update not synced');
    }
  }

  static async getUserStats() {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/stats`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available for stats');
    }
    
    return {
      total_goals: 3,
      achieved_goals: 0,
      achievement_rate: 0
    };
  }

  // Drink History
  static async getDrinkHistory(limit?: number) {
    try {
      const url = limit ? `${API_BASE_URL}/user/${this.userId}/drinks?limit=${limit}` : `${API_BASE_URL}/user/${this.userId}/drinks`;
      const response = await fetch(url);
      if (response.ok) {
        const data = await response.json();
        return data.drinks;
      }
    } catch (error) {
      console.warn('Backend not available for drink history');
    }
    
    // Fallback to localStorage
    const stored = localStorage.getItem('drinkHistory');
    return stored ? JSON.parse(stored) : [];
  }

  static async getTodayDrinks() {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/drinks/today`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available for today drinks');
    }
    
    // Fallback to localStorage with date filtering
    const stored = localStorage.getItem('drinkHistory');
    if (stored) {
      const drinks = JSON.parse(stored);
      const today = new Date().toDateString();
      return {
        drinks: drinks.filter((drink: any) => new Date(drink.timestamp).toDateString() === today),
        totals: { calories: 0, sugar_g: 0, caffeine_mg: 0, water_ml: 0, drink_count: 0 }
      };
    }
    
    return { drinks: [], totals: { calories: 0, sugar_g: 0, caffeine_mg: 0, water_ml: 0, drink_count: 0 } };
  }

  static async getWeeklyStats() {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/drinks/weekly-stats`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available for weekly stats');
    }
    
    return {
      weekly_stats: {
        total_drinks: 0,
        total_calories: 0,
        avg_drinks_per_day: 0,
        most_consumed_drink: 'None'
      }
    };
  }

  static async getHealthInsights() {
    try {
      const response = await fetch(`${API_BASE_URL}/user/${this.userId}/health-insights`);
      if (response.ok) {
        return await response.json();
      }
    } catch (error) {
      console.warn('Backend not available for health insights');
    }
    
    return { insights: ['Keep tracking your drinks for personalized insights!'] };
  }
}